### ISSUE TITLE
Easy to break migration: adding new property in one version, then changing property type in later version

### ISSUE THREAD

**Issue by NeverwinterMoon** [realm/realm-cocoa#6263]
## Goals
Let's say I had this:
```
/// schema version 0
class Client: Object {
  @objc dynamic var name: String = ""
}
```
Later on
```
/// schema version 1
class Client: Object {
  let id = RealmOptional<Int>()
  @objc dynamic var name: String = ""
}
```

Because the new property is added, no migration step is necessary, as far I as know, just bumping the schema version is enough.

Later on
```
/// schema version 2
class Client: Object {
  @objc dynamic var id: Int = 0
  @objc dynamic var name: String = ""
}

/// MigrationBlock
if oldSchemaVersion < 1 {
        print("Migrating to 1")
}

if oldSchemaVersion < 2 {
  migration.enumerateObjects(ofType: Client.className()) { oldObject, newObject in
    let id = oldObject!["id"] as? Int
    newObject!["id"] = id ?? 0
  }
}
```

## Expected Results
Migration works from 0 to 1. From 1 to 2. And from 0 to 2.

## Actual Results
If the client is migrating from 0 to 1. Then from 1 to 2, everything works as expected.

But if the client is migrating from 0 to 2, an exception is thrown `Invalid property name 'id' for class 'Client'.

I understand why this happens - when migrating to schema version 2 we assume that `oldObject!["id"]` exists but because it didn't exist in schema version 0, it crashes. But this is very easy to miss, especially if migration from 1 to 2 done after some time. The documentation example says:

```
// Nothing to do!
// Realm will automatically detect new properties and removed properties
// And will update the schema on disk automatically
```

So it's a logical thing to do when migrating to version 1 - nothing. But if later on the newly added property's type is changed, the hell breaks loose.

Right now my solution was to:
```
let id: Int?
if oldObject?.objectSchema.properties.contains(where: { $0.name == "id" }) ?? false {
  id = oldObject!["id"] as? Int
} else {
  id = nil
}
newObject!["id"] = id ?? 0
```

My point is, such a case is too easy to miss and maybe it makes sense to at least document this pitfall.

**Comment by winstondu** [comment]
Agreed. The realm migration model just does not sustain certain quirks of linear migration.

**Comment by DominicFrei** [comment]
I created #7069 to better explain what's going on here and how this should be handled.

TL;DR
The problem is that the migration block does not know about old schemas and can therefore not know if the property existed before and which type it had since the schema only reflects the current properties and types thereof.
This has to be reflected in the migration block. The developer knows about previous migrations and schemas and has to take into account jumps from 0->1, 1->2 as well as 0->2.

This issue is to be addressed with an updated example as mentioned above.

**Comment by duncangroenewald** [comment]
Not sure if I missed this in the docs on schema updates for synced realms but how do these migrations occur for synced realms where the server schema is updated, say with new properties, and the client is either:
a) running with an old version of the schema, 
b) just updated to the new version of the schema but not run been run yet, or 
c) has an old version of the schema but is not currently running yet.

**Comment by DominicFrei** [comment]
@simonask and @tkaye407 can maybe shed some light on the sync part.
I'm not too familiar with that area yet.

**Comment by ianpward** [comment]
@duncangroenewald the docs are here - 
https://docs.mongodb.com/realm/ios/migrations#synced-migration

**Comment by duncangroenewald** [comment]
@ianpward - does that apply to MongoDB Realm ?  I was under the impression that with MongoDB Realm you cannot just change the schema on the client as this will prevent the record from syncing with MongoDB if the schema does not match the defined schema in the MongoDB Realm App UNLESS the Realm App is running in Developer Mode.

**Comment by ianpward** [comment]
Yes developer mode the client can make changes to the sync schema. 

For production mode you make changes on the server side first.

**Comment by duncangroenewald** [comment]
So if you modify the schema on the server how do you coordinate that change with all the clients ?  What happens if a client has an older version of the schema and tries to modify some data ?  Won't the sync fail because the record schema does not match the server schema ?

**Comment by ianpward** [comment]
@duncangroenewald The answers to your questions are in the docs I sent, is it possible for you to review them?
https://docs.mongodb.com/realm/ios/migrations#synced-migration

Also, this has has diverged from OP so please open a forums post if you have questions after reviewing the docs

**Comment by duncangroenewald** [comment]
I didn't see any mention of basic migration steps for a synced realm such as updating the Realm App schema on the server.  In the RealmDB Sync documents(https://docs.mongodb.com/realm/sync/configure-your-data-model) it mentions:

`While Development Mode is on, MongoDB Realm doesn't validate writes against your data model, allowing you to freely update your Realm Object Model. When you turn off Development Mode, MongoDB Realm automatically updates your Realm Schema and starts to **enforce data validation** for your Atlas cluster based on it.`

So if the schema is modified to add a new property and a new client using an old version of the model connects and starts writing data what is supposed to happen ?  Will the client records fail data validation and therefor fail to sync because they are missing a property ?

For now we will just run a bunch of tests to see what actually happens and based on that make some suggestions to improve the docs if required.

**Comment by ianpward** [comment]
@duncangroenewald Tag me on a forums post with your questions and I will be happy to answer

**Comment by DominicFrei** [comment]
> I created #7069 to better explain what's going on here and how this should be handled.
> 
> TL;DR
> The problem is that the migration block does not know about old schemas and can therefore not know if the property existed before and which type it had since the schema only reflects the current properties and types thereof.
> This has to be reflected in the migration block. The developer knows about previous migrations and schemas and has to take into account jumps from 0->1, 1->2 as well as 0->2.
> 
> This issue is to be addressed with an updated example as mentioned above.

The [above mentioned PR](https://github.com/realm/realm-cocoa/pull/7077) is ready for review now.

@NeverwinterMoon @winstondu Feel free to have a look at that and tell me if I need to further clarify.

@duncangroenewald I only just realised that you're having a different issue. Would it make sense to open another issue for that so that those two do not get mixed up (also, this issue might get closed soon with the PR)?

**Comment by DominicFrei** [comment]
I am going to close this issue for now since there was no further reply.

Is there are still open questions feel free to re-open it.
